searchlib/src/vespa/searchlib/tensor/nearest_neighbor_index.h

```diff
@@ -36,6 +38,38 @@ class NearestNeighborIndex {
     };
     virtual ~NearestNeighborIndex() {}
     virtual void add_document(uint32_t docid) = 0;
+
+    /**
+     * Performs the prepare step in a two-phase operation to add a document to the index.
+     *
+     * This function can be called by any thread.
+     * The document to add is represented by the given vector as it is _not_ stored in the enclosing tensor attribute at this point in time.
+     * It should return the result of the costly and non-modifying part of this operation.
+     * The given read guard must be kept in the result.
+     */
+    virtual std::unique_ptr<PrepareResult> prepare_add_document(uint32_t docid,
+                                                                vespalib::tensor::TypedCells vector,
+                                                                vespalib::GenerationHandler::Guard read_guard) const {
+        // TODO: Make it pure virtual after more wiring is complete.
+        (void) docid;
+        (void) vector;
+        (void) read_guard;
+        return std::unique_ptr<PrepareResult>();
+    }
+
+    /**
+     * Performs the complete step in a two-phase operation to add a document to the index.
+     *
+     * This function is only called by the attribute writer thread.
+     * It must wait for the result from the prepare step (via the future) before it does the modifying changes.
+     */
+    virtual void complete_add_document(uint32_t docid,
+                                       std::future<std::unique_ptr<PrepareResult>> prepare_result) {
```

Comment:
promise and future handling should be at a layer above this code. Consider using `std::unique_ptr<PrepareResult>`or `PrepareResult&`instead of `std::future<std::unique_ptr<PrepareResult>>`.

Reply:
Fixed